tracer/src/Datadog.Trace/Configuration/TracerSettingsSnapshot.cs

```diff
@@ -0,0 +1,328 @@
+        LogsInjectionEnabled = settings.LogSubmissionSettings.LogsInjectionEnabled;
+        GlobalSamplingRate = settings.GlobalSamplingRateInternal;
+        DisabledIntegrationNames = GetHashSet(settings.DisabledIntegrationNamesInternal);
+        // this is the easiest way to create a "snapshot" of the settings
+        Integrations = new ImmutableIntegrationSettingsCollection(settings.IntegrationsInternal, EmptyHashSet);
+        GlobalTags = GetDictionary(settings.GlobalTagsInternal);
+        HeaderTags = GetDictionary(settings.HeaderTagsInternal);
+        GrpcTags = GetDictionary(settings.GrpcTagsInternal);
+        TracerMetricsEnabled = settings.TracerMetricsEnabledInternal;
+        StatsComputationEnabled = settings.StatsComputationEnabledInternal;
+        KafkaCreateConsumerScopeEnabled = settings.KafkaCreateConsumerScopeEnabledInternal;
+        StartupDiagnosticLogEnabled = settings.StartupDiagnosticLogEnabledInternal;
+        DirectLogSubmissionBatchPeriod = settings.LogSubmissionSettings.DirectLogSubmissionBatchPeriod;
+    }
+
+    public string? Environment { get; }
+
+    public string? ServiceName { get; }
+
+    public string? ServiceVersion { get; }
+
+    public bool TraceEnabled { get; }
+
+    public ExporterSettingsSnapshot Exporter { get; }
+
+    public bool AnalyticsEnabled { get; }
+
+    public int MaxTracesSubmittedPerSecond { get; }
+
+    public string? CustomSamplingRules { get; }
+
+    public bool? LogsInjectionEnabled { get; }
+
+    public double? GlobalSamplingRate { get; }
+
+    public HashSet<string>? DisabledIntegrationNames { get; }
+
+    public ImmutableIntegrationSettingsCollection Integrations { get; }
+
+    public IDictionary<string, string>? GlobalTags { get; }
+
+    public IDictionary<string, string>? HeaderTags { get; }
+
+    public IDictionary<string, string>? GrpcTags { get; }
+
+    public bool TracerMetricsEnabled { get; }
+
+    public bool StatsComputationEnabled { get; }
+
+    public bool KafkaCreateConsumerScopeEnabled { get; }
+
+    public bool StartupDiagnosticLogEnabled { get; }
+
+    public TimeSpan DirectLogSubmissionBatchPeriod { get; }
+
+    private static Dictionary<string, string>? GetDictionary(IDictionary<string, string>? source)
+    {
+        if (source is null or { Count: 0 })
+        {
+            return null;
+        }
+
+        return new Dictionary<string, string>(source);
+    }
+
+    private static HashSet<string>? GetHashSet(HashSet<string>? source)
+    {
+        if (source is null or { Count: 0 })
+        {
+            return null;
+        }
+
+        return new HashSet<string>(source);
+    }
+
+    /// <summary>
+    /// Record any changes to the provided <see cref="TracerSettings"/> in <paramref name="telemetry"/>
+    /// </summary>
+    public void RecordChanges(TracerSettings settings, IConfigurationTelemetry telemetry)
+    {
+        RecordIfChanged(telemetry, ConfigurationKeys.Environment, Environment, settings.EnvironmentInternal);
```

Comment:
Honestly I thought you would source-generate this part 😛 

If you use ImmutableTracerSettings as snapshot, and source-generate a method that returns all the properties as IEnumerable, I think we can remove most of the boilerplate code and this way we don't forget to update it when adding a new property

Reply:
Yeah, it's a trade off. We allocate more if we do it that way 🤷‍♂️ I'm not passionate about it. But frankly this is easier right not 😉